# ViReMaShiny
 A Shiny app that generates figures for ViReMa analyses

## What is ViReMa?
ViReMa is an algorithm developed by the Routh and Johnson Labs providing “a versatile platform for rapid, sensitive and nucleotide-resolution detection of recombination junctions in viral genomes using next-generation sequencing data”. Code for ViReMa is written in Python and can be downloaded [here](https://bioconda.github.io/recipes/virema/README.html).

## Documentation for ViReMaShiny
[Tutorial and Vignettes](https://jayeung12.github.io/)


[Preprint]()

## Relevant Papers
The most recent ViReMa paper:
Sotcheff S, Zhou Y, Sun Y, Johnson JE, Torbett B, Routh AL. ViReMa: A Virus Recombination Mapper of Next-Generation Sequencing data characterizes diverse recombinant viral nucleic acids. bioRxiv. 2022:2022.03.12.484090. doi:10.1101/2022.03.12.484090


The original ViReMa paper:
Andrew Routh, John E. Johnson, Discovery of functional genomic motifs in viruses with ViReMa-a Virus Recombination Mapper-for analysis of next-generation sequencing data, Nucleic Acids Research, Volume 42, Issue 2, 1 January 2014, Page e11, https://doi.org/10.1093/nar/gkt916
